In [1]:
import pandas as pd
import numpy as np
import datetime

# Data Extraction

This document is related on the extraction of the data, focused on the time period from 01-01-200 to 09-01-2020. In order to work on this it's needed a database with closing prices of the american market, i our case a csv file.

## Reading & Cleaning Data

In [2]:
'''Read a dataframe which only contains the (tickers,company, company sector)'''
df_sectors = pd.read_csv('../data/raw/sp_sectors.csv', skiprows=None)
df_sectors

Ticker                            Company                  Sector
0      AAP            ADVANCE AUTO PARTS, INC  Consumer Discretionary
1     AMZN                    AMAZON.COM ,INC  Consumer Discretionary
2     APTV                          APTIV PLC  Consumer Discretionary
3      AZO                     AUTOZONE, INC.  Consumer Discretionary
4      BBY                  BEST BUY CO, INC.  Consumer Discretionary
..     ...                                ...                     ...
500    PXD  PIONEER NATURAL RESOURCES COMPANY                  Energy
501    SLB               SCHLUMBERGER LIMITED                  Energy
502    VLO          VALERO ENERGY CORPORATION                  Energy
503    WMB           WILLIAMS COMPANIES, INC.                  Energy
504    XOM            EXXON MOBIL CORPORATION                  Energy

[505 rows x 3 columns]

In [3]:
'''Read the financial data'''
df = pd.read_csv('../data/raw/data.csv', header=None)
df

0           1      2      3      4      5           6    7    8   \
0           A  1999-11-18  45.50  50.00  40.00  44.00  44739900.0  0.0  1.0   
1           A  1999-11-19  42.94  43.00  39.81  40.38  10897100.0  0.0  1.0   
2           A  1999-11-22  41.31  44.00  40.06  44.00   4705200.0  0.0  1.0   
3           A  1999-11-23  42.50  43.63  40.25  40.25   4274400.0  0.0  1.0   
4           A  1999-11-24  40.13  41.94  40.00  41.06   3464400.0  0.0  1.0   
...       ...         ...    ...    ...    ...    ...         ...  ...  ...   
37594096  ZZZ  2017-08-29   0.01   0.01   0.01   0.01         0.0  0.0  1.0   
37594097  ZZZ  2017-08-30   0.01   0.01   0.01   0.01         0.0  0.0  1.0   
37594098  ZZZ  2017-08-31   0.01   0.01   0.01   0.01         0.0  0.0  1.0   
37594099  ZZZ  2017-09-01   0.01   0.01   0.01   0.01         0.0  0.0  1.0   
37594100  ZZZ  2017-09-05   0.01   0.01   0.01   0.01       100.0  0.0  1.0   

                 9          10         11         12          13  
0         30.238319  33.228922  26.583138  29.241452  44739900.0  
1         28.536998  28.576873  26.456868  26.835678  10897100.0  
2         27.453736  29.241452  26.623012  29.241452   4705200.0  
3         28.244584  28.995557  26.749282  26.749282   4274400.0  
4         26.669533  27.872420  26.583138  27.287591   3464400.0  
...             ...        ...        ...        ...         ...  
37594096   0.010000   0.010000   0.010000   0.010000         0.0  
37594097   0.010000   0.010000   0.010000   0.010000         0.0  
37594098   0.010000   0.010000   0.010000   0.010000         0.0  
37594099   0.010000   0.010000   0.010000   0.010000         0.0  
37594100   0.010000   0.010000   0.010000   0.010000       100.0  

[37594101 rows x 14 columns]

In [4]:
'''Make a list out of the tickers'''
ticker_list= df_sectors['Ticker'].values.tolist()
ticker_list.sort()
print(ticker_list, len(ticker_list))

['A', 'AAL', 'AAP', 'AAPL', 'ABBV', 'ABC', 'ABMD', 'ABT', 'ACN', 'ADBE', 'ADI', 'ADM', 'ADP', 'ADSK', 'AEE', 'AEP', 'AES', 'AFL', 'AIG', 'AIV', 'AIZ', 'AJG', 'AKAM', 'ALB', 'ALGN', 'ALK', 'ALL', 'ALLE', 'ALXN', 'AMAT', 'AMCR', 'AMD', 'AME', 'AMGN', 'AMP', 'AMT', 'AMZN', 'ANET', 'ANSS', 'ANTM', 'AON', 'AOS', 'APA', 'APD', 'APH', 'APTV', 'ARE', 'ATO', 'ATVI', 'AVB', 'AVGO', 'AVY', 'AWK', 'AXP', 'AZO', 'BA', 'BAC', 'BAX', 'BBY', 'BDX', 'BEN', 'BF_B', 'BIIB', 'BIO', 'BK', 'BKNG', 'BKR', 'BLK', 'BLL', 'BMY', 'BR', 'BRK_B', 'BSX', 'BWA', 'BXP', 'C', 'CAG', 'CAH', 'CARR', 'CAT', 'CB', 'CBOE', 'CBRE', 'CCI', 'CCL', 'CDNS', 'CDW', 'CE', 'CERN', 'CF', 'CFG', 'CHD', 'CHRW', 'CHTR', 'CI', 'CINF', 'CL', 'CLX', 'CMA', 'CMCSA', 'CME', 'CMG', 'CMI', 'CMS', 'CNC', 'CNP', 'COF', 'COG', 'COO', 'COP', 'COST', 'CPB', 'CPRT', 'CRM', 'CSCO', 'CSX', 'CTAS', 'CTLT', 'CTSH', 'CTVA', 'CTXS', 'CVS', 'CVX', 'CXO', 'D', 'DAL', 'DD', 'DE', 'DFS', 'DG', 'DGX', 'DHI', 'DHR', 'DIS', 'DISCA', 'DISCK', 'DISH', 'DLR', 'DL

In [5]:
'''Eliminate the columns that are irrelevant for the study'''
df_cleaned = df.drop([2,3,4,6,7,8,9,10,11,12,13], axis=1)

'''
Filter the data so that we maintain only information about tickers
that conforms the index, we do this by using the ticker list we aleady made
'''
df_cleaned = df_cleaned.loc[df_cleaned[0].isin(ticker_list)]
df_cleaned

0           1       5
0           A  1999-11-18   44.00
1           A  1999-11-19   40.38
2           A  1999-11-22   44.00
3           A  1999-11-23   40.25
4           A  1999-11-24   41.06
...       ...         ...     ...
37579467  ZTS  2020-11-16  165.29
37579468  ZTS  2020-11-17  165.43
37579469  ZTS  2020-11-18  163.50
37579470  ZTS  2020-11-19  166.31
37579471  ZTS  2020-11-20  166.20

[3581566 rows x 3 columns]

In [6]:
'''Check what type we have on each column'''
df_cleaned.dtypes

0     object
1     object
5    float64
dtype: object

In [7]:
'''Change the "dates" from object to datetime64'''
df_cleaned[1] = pd.to_datetime(df_cleaned[1])
df_cleaned.dtypes

0            object
1    datetime64[ns]
5           float64
dtype: object

In [8]:
'''
Filter the financial dataframe with the time period we would like
to investigate; in this case: from 01-01-2000 to 09-01-2020
'''
initial_date = datetime.datetime(2000, 1, 1)
final_date = datetime.datetime(2020, 9, 1)

initial_limit = df_cleaned[1] >= initial_date
final_limit = df_cleaned[1] <= final_date

'''
With all done, before giving the dataframe a matrix format,
it is important to get the right format to apply some pandas functions
'''
df_cleaned = df_cleaned[initial_limit & final_limit]
df_cleaned.columns = ['Ticker', 'Date', 'Closing Price']
df_cleaned.reset_index(inplace=True,drop=True)
df_cleaned

Ticker       Date  Closing Price
0            A 2000-01-03          72.00
1            A 2000-01-04          66.50
2            A 2000-01-05          61.56
3            A 2000-01-06          60.00
4            A 2000-01-07          65.00
...        ...        ...            ...
2321546    ZTS 2020-08-26         159.76
2321547    ZTS 2020-08-27         159.61
2321548    ZTS 2020-08-28         160.00
2321549    ZTS 2020-08-31         160.10
2321550    ZTS 2020-09-01         162.05

[2321551 rows x 3 columns]

**Note**. It is important to enphazise that one ticker will be missed from here. That's because the ticker 'VNT' starts to show data since 10/09/2020, and that time is out of the time period we focus on. So we are left with 504 tickers.

### Comparison

In [9]:
'''
If we search in the original dataframe, we will see we have registers after
the final date of our time time period
'''
df[df[0] == 'VNT'].head(5)

0           1      2      3        4      5           6    7    8   \
35649209  VNT  2020-10-09  32.50  32.90  27.8500  28.11  19985326.0  0.0  1.0   
35649210  VNT  2020-10-12  28.27  29.00  27.6200  27.85   6532325.0  0.0  1.0   
35649211  VNT  2020-10-13  28.29  28.42  26.3636  26.76   4937341.0  0.0  1.0   
35649212  VNT  2020-10-14  26.80  28.20  26.4300  28.01   7992242.0  0.0  1.0   
35649213  VNT  2020-10-15  27.34  28.11  26.5100  27.75   7194620.0  0.0  1.0   

             9      10       11     12          13  
35649209  32.50  32.90  27.8500  28.11  19985326.0  
35649210  28.27  29.00  27.6200  27.85   6532325.0  
35649211  28.29  28.42  26.3636  26.76   4937341.0  
35649212  26.80  28.20  26.4300  28.01   7992242.0  
35649213  27.34  28.11  26.5100  27.75   7194620.0

In [10]:
'''
If we search in the filtered dataframe, we will see we have no registers
because of our selection of time period of analysis
'''
df_cleaned[df_cleaned['Ticker'] == 'VNT']

Empty DataFrame
Columns: [Ticker, Date, Closing Price]
Index: []

## Creating the Data Matrix

With the following functions we create a data matrix with dimensions ($N\times t$), where $N$ is the number of tickers and $t$ is the numbers of trading days in the U.S. Stock Market

In [11]:
'''
We take the dataframe and use the columns ("Ticker", Dates) to be the respectives 
rows and columns of the data matrix, with the values being the "Closing Prices"
'''
df_data = df_cleaned.pivot(index='Ticker', columns='Date')['Closing Price']
df_data.columns = df_data.columns.strftime('%x')
df_data.reset_index(inplace=True)
df_data

Date Ticker  01/03/00  01/04/00  01/05/00  01/06/00  01/07/00  01/10/00  \
0         A     72.00     66.50     61.56     60.00     65.00     68.94   
1       AAL       NaN       NaN       NaN       NaN       NaN       NaN   
2       AAP       NaN       NaN       NaN       NaN       NaN       NaN   
3      AAPL    111.94    102.50    104.00     95.00     99.50     97.75   
4      ABBV       NaN       NaN       NaN       NaN       NaN       NaN   
..      ...       ...       ...       ...       ...       ...       ...   
499     YUM     37.31     36.56     36.81     36.44     35.63     37.06   
500     ZBH       NaN       NaN       NaN       NaN       NaN       NaN   
501    ZBRA     56.31     55.50     56.75     53.50     52.91     54.69   
502    ZION     55.50     52.81     53.06     53.50     53.63     52.66   
503     ZTS       NaN       NaN       NaN       NaN       NaN       NaN   

Date  01/11/00  01/12/00  01/13/00  ...  08/19/20  08/20/20  08/21/20  \
0        68.00     66.63     67.63  ...     97.96     97.78     98.30   
1          NaN       NaN       NaN  ...     12.68     12.50     12.16   
2          NaN       NaN       NaN  ...    160.81    157.44    156.11   
3        92.75     87.19     96.75  ...    462.83    473.10    497.48   
4          NaN       NaN       NaN  ...     96.10     95.20     94.86   
..         ...       ...       ...  ...       ...       ...       ...   
499      36.81     36.44     37.00  ...     93.93     95.37     96.10   
500        NaN       NaN       NaN  ...    137.25    136.80    136.16   
501      55.06     53.69     53.88  ...    283.13    276.55    275.33   
502      49.88     50.31     54.69  ...     32.91     32.13     31.59   
503        NaN       NaN       NaN  ...    158.58    158.54    158.37   

Date  08/24/20  08/25/20  08/26/20  08/27/20  08/28/20  08/31/20  09/01/20  
0        97.60     98.34    99.330     99.02     99.90    100.42   101.120  
1        13.44     13.14    12.785     13.28     13.59     13.05    13.100  
2       157.14    156.06   156.480    154.72    154.77    156.31   156.390  
3       503.43    499.30   506.090    500.04    499.23    129.04   134.180  
4        94.51     94.06    94.350     94.30     94.17     95.77    92.240  
..         ...       ...       ...       ...       ...       ...       ...  
499      96.56     96.48    95.330     95.47     97.45     95.85    96.150  
500     136.81    137.60   136.830    140.53    141.67    140.88   140.480  
501     275.39    277.86   284.740    277.26    280.16    286.53   287.160  
502      33.04     33.06    32.280     33.29     33.02     32.16    32.185  
503     156.76    158.58   159.760    159.61    160.00    160.10   162.050  

[504 rows x 5201 columns]

In [12]:
'''Create an list with the correct order of the tickers per sector'''
tickers_sorted = df_sectors['Ticker'].values.tolist()
print(tickers_sorted)

['AAP', 'AMZN', 'APTV', 'AZO', 'BBY', 'BKNG', 'BWA', 'CCL', 'CMG', 'DG', 'DHI', 'DLTR', 'DPZ', 'DRI', 'EBAY', 'ETSY', 'EXPE', 'F', 'GM', 'GPC', 'GPS', 'GRMN', 'HAS', 'HBI', 'HD', 'HLT', 'KMX', 'LB', 'LEG', 'LEN', 'LKQ', 'LOW', 'LVS', 'MAR', 'MCD', 'MGM', 'MHK', 'NCLH', 'NKE', 'NVR', 'NWL', 'ORLY', 'PHM', 'POOL', 'PVH', 'RCL', 'RL', 'ROST', 'SBUX', 'TGT', 'TIF', 'TJX', 'TPR', 'TSCO', 'UA', 'UAA', 'ULTA', 'VFC', 'WHR', 'WYNN', 'YUM', 'ADM', 'BF_B', 'CAG', 'CHD', 'CL', 'CLX', 'COST', 'CPB', 'EL', 'GIS', 'HRL', 'HSY', 'K', 'KHC', 'KMB', 'KO', 'KR', 'LW', 'MDLZ', 'MKC', 'MNST', 'MO', 'PEP', 'PG', 'PM', 'SJM', 'STZ', 'SYY', 'TAP', 'TSN', 'WBA', 'WMT', 'A', 'ABBV', 'ABC', 'ABMD', 'ABT', 'ALGN', 'ALXN', 'AMGN', 'ANTM', 'BAX', 'BDX', 'BIIB', 'BIO', 'BMY', 'BSX', 'CAH', 'CERN', 'CI', 'CNC', 'COO', 'CTLT', 'CVS', 'DGX', 'DHR', 'DVA', 'DXCM', 'EW', 'GILD', 'HCA', 'HOLX', 'HSIC', 'HUM', 'IDXX', 'ILMN', 'INCY', 'IQV', 'ISRG', 'JNJ', 'LH', 'LLY', 'MCK', 'MDT', 'MRK', 'MTD', 'MYL', 'PFE', 'PKI', 'PRGO

In [13]:
'''Sort the rows in the matrix acording to each sector'''
df_data.Ticker = df_data.Ticker.astype('category')
df_data.Ticker.cat.set_categories(tickers_sorted, inplace=True)
df_data = df_data.sort_values(['Ticker'])
df_data

Date Ticker  01/03/00  01/04/00  01/05/00  01/06/00  01/07/00  01/10/00  \
2       AAP       NaN       NaN       NaN       NaN       NaN       NaN   
36     AMZN    89.380    81.940    71.750     65.56     69.56     69.19   
45     APTV       NaN       NaN       NaN       NaN       NaN       NaN   
54      AZO    30.560    30.440    30.060     29.06     30.44     30.69   
58      BBY    57.500    57.380    56.690     56.50     58.25     57.63   
..      ...       ...       ...       ...       ...       ...       ...   
384     PXD     8.438     8.563     8.563      8.50      8.75      8.50   
410     SLB    54.940    54.190    54.880     57.75     58.81     59.38   
467     VLO    19.130    18.940    19.810     20.25     20.00     19.94   
485     WMB    31.000    30.310    32.000     32.88     33.75     33.13   
495     XOM    78.310    76.810    80.630     85.19     84.94     83.75   

Date  01/11/00  01/12/00  01/13/00  ...  08/19/20  08/20/20  08/21/20  \
2          NaN       NaN       NaN  ...    160.81    157.44    156.11   
36      66.750    63.560    65.940  ...   3260.48   3297.37   3284.72   
45         NaN       NaN       NaN  ...     85.42     84.49     83.88   
54      30.060    29.810    30.000  ...   1210.00   1203.50   1209.07   
58      55.880    54.440    53.000  ...    111.96    111.70    114.00   
..         ...       ...       ...  ...       ...       ...       ...   
384      8.625     8.188     8.438  ...    104.77    102.67    102.11   
410     60.690    59.190    60.880  ...     19.72     18.77     18.58   
467     19.630    19.750    19.940  ...     54.26     52.70     52.62   
485     32.750    32.310    33.060  ...     21.65     21.40     21.41   
495     84.750    83.500    85.440  ...     41.96     41.32     41.01   

Date  08/24/20  08/25/20  08/26/20  08/27/20  08/28/20  08/31/20  09/01/20  
2       157.14    156.06    156.48    154.72    154.77    156.31    156.39  
36     3307.46   3346.49   3441.85   3400.00   3401.80   3450.96   3499.12  
45       85.50     85.36     85.44     85.34     87.25     86.12     87.05  
54     1209.93   1201.66   1198.01   1195.89   1189.12   1196.31   1222.43  
58      117.37    112.64    112.31    111.22    111.23    110.91    113.50  
..         ...       ...       ...       ...       ...       ...       ...  
384     104.75    106.92    104.41    105.38    105.60    103.93    102.98  
410      19.55     19.26     18.67     18.70     19.42     19.01     18.63  
467      54.99     54.25     53.70     53.44     54.89     52.59     52.47  
485      21.68     21.39     21.05     20.96     21.07     20.76     20.96  
495      42.22     40.88     40.01     39.74     40.69     39.94     39.43  

[504 rows x 5201 columns]

**Note**. This is not the final matrix. Now, we'll eliminate the rows which has 'NaN' values. The easiest way to do that is to eliminate those which has a 'NaN' value the first day, that is: 01/03/2000.

In [14]:
'''Count how many rows have any "NaN" values'''
df_data['01/03/00'].isnull().values.ravel().sum()

130

* This means we have to eliminate 130 rows from the dataframe.
* This leave us with 374 ticker rows.

In [15]:
'''Create a list with the rows index we are about to eliminate'''
drop_index_list = df_data[df_data['01/03/00'].isnull()].index.tolist()
print(drop_index_list, len(drop_index_list))

[2, 45, 101, 129, 141, 166, 171, 202, 208, 214, 221, 279, 288, 325, 440, 451, 452, 456, 492, 262, 289, 300, 371, 4, 24, 39, 104, 117, 149, 168, 215, 237, 245, 248, 500, 503, 1, 27, 78, 125, 177, 195, 220, 231, 239, 246, 277, 333, 353, 429, 445, 453, 470, 498, 8, 37, 50, 70, 86, 113, 148, 182, 187, 194, 206, 224, 244, 261, 273, 292, 336, 355, 385, 387, 420, 431, 463, 490, 30, 87, 89, 119, 126, 140, 290, 368, 488, 82, 137, 160, 172, 93, 134, 135, 176, 203, 204, 291, 329, 345, 346, 439, 447, 466, 52, 162, 337, 20, 34, 81, 90, 100, 128, 191, 233, 302, 306, 317, 326, 363, 379, 416, 424, 483, 123, 174, 193, 266, 313, 381] 130


In [16]:
'''Eliminate rows we do not need'''
df_data = df_data.drop(drop_index_list, axis=0)

'''Save the resultant data matrix'''
df_data.set_index('Ticker')
df_data.to_csv('../data/interim/data_matrix.csv', index=False) 

## Sectors 

Now, just to see the perspective, let's see how each sector has changed.

We have to modify the file with the sectors info. It has 505 tickers, so we must drop the same rows as the data matrix PLUS the ticker that wasn't even in the data matrix, i.e., 'VNT' (index 295).

In [17]:
'''Add the last ticker in a new index list to drop'''
new_drop_index_list = drop_index_list + [295]
new_drop_index_list.sort()

'''Eliminate the ticker we do not need'''
df_sectors_cleaned = df_sectors.drop(new_drop_index_list, axis=0)

'''Save the remaining tickers with their sector in a csv file'''
df_sectors_cleaned
df_sectors_cleaned.to_csv('../data/interim/sp_sectors_filtered.csv', index=False)

In [18]:
'''Create a list with the name of the sectors, a total of 11'''
sectors_list = list(df_sectors['Sector'].drop_duplicates(keep='first'))

'''Here we count how much part of each sector was removed'''
df_sectors_diff = pd.DataFrame(index=sectors_list)
df_sectors_diff.index.name = 'Sectors of the Market'
df_sectors_diff['Before'] = pd.Series(df_sectors['Sector'].value_counts())
df_sectors_diff['After'] = pd.Series(df_sectors_cleaned['Sector'].value_counts())
df_sectors_diff['Difference'] = df_sectors_diff['Before'] - df_sectors_diff['After']
df_sectors_diff

Before  After  Difference
Sectors of the Market                            
Consumer Discretionary      61     47          14
Consumer Staples            32     24           8
Health Care                 63     44          19
Industrials                 73     48          25
Information Technology      73     52          21
Materials                   28     21           7
Real Estate                 31     24           7
Communication Services      26     20           6
Utilities                   28     27           1
Financials                  65     49          16
Energy                      25     18           7